pkg/virt-launcher/virtwrap/converter/converter.go

```diff
@@ -325,6 +333,50 @@ func SetOptimalIOMode(disk *api.Disk) error {
 	return nil
 }
 
+// SetOptimalBlockIO sets the BlockIO fields to match the backing device
+func SetOptimalBlockIO(disk *api.Disk) error {
+	if disk.BlockIO != nil {
+		return nil
+	}
+
+	var path string
+	if disk.Source.File != "" {
+		path = disk.Source.File
+	} else if disk.Source.Dev != "" {
+		path = disk.Source.Dev
+	} else {
+		return nil
+	}
+
+	f, err := os.OpenFile(path, syscall.O_RDONLY, 0)
+	if err != nil {
+		return nil
+	}
+	logicalSize, err := unix.IoctlGetInt(int(f.Fd()), unix.BLKBSZGET)
+	if err != nil {
+		return nil
+	}
+	physicalSize, err := unix.IoctlGetInt(int(f.Fd()), unix.BLKSSZGET)
+	if err != nil {
+		return nil
+	}
+	defer util.CloseIOAndCheckErr(f, nil)
+
+	if logicalSize <= 0 && physicalSize <= 0 {
+		return nil
+	}
+	if logicalSize <= 0 || physicalSize <= 0 {
+		if logicalSize > physicalSize {
+			physicalSize = logicalSize
+		} else {
+			logicalSize = physicalSize
+		}
+	}
+	disk.BlockIO.LogicalBlockSize = logicalSize
+	disk.BlockIO.PhysicalBlockSize = physicalSize
```

Comment:
since we haven't been defining these values in the past, one concern I have is about consistency for VM restarts after kubevirt would be upgraded with this new behavior. It's possible the physical/logical block sizes within the VM guest may differ from before the restart due to us auto detecting this on the host now.

I'm not sure we'll be able to definitively say this change in behavior won't cause issues for someone. 

@Omar007 i think we should consider figuring out a way to make this behavior opt-in. 

One way would be to change up the way this is exposed on the VirtualMachineInstance spec.

```
type CustomBlockSize struct {
	LogicalBlockSize  int `json:"logicalBlockSize,omitempty"`
	PhysicalBlockSize int `json:"physicalBlockSize,omitempty"`
}

// Only one can be picked from this struct.
Type BlockSize {
    MatchVolumeBlockSize *FeatureState `json:"matchVolumeBlockSize,omitempty"`
    CustomBlockSize *CustomBlockSize `json:"customBlockSize,omitempty"`
}
```

That would allow someone to opt in to the auto detetion logic and also have the ability to set their own custom values. thoughts?

Reply:
> I'm not sure we'll be able to definitively say this change in behavior won't cause issues for someone.
Same. While there are cases where it can be fine, it is not something I feel confident with we should just have happen by itself. 

> i think we should consider figuring out a way to make this behavior opt-in.
We should. That's why I wasn't sure about having the detection as a part of this MR or thought to at least put it behind a feature flag or something but I like this option as well (more, actually, since you can then do it per-VM). 
Regardless of the final solution/implementation, in my opinion, the default should be what it is now (which is up to QEMU but I doubt they'll suddenly change the default from 512n to something else) to prevent breaking existing deployments/VMs.

Reply:
I've pushed some changes that take this idea.